### DISCUSSION TITLE
How to understand shader variants in MaterialBuilder

### DISCUSSION THREAD

**Discussion by SC5Shout** [google/filament#7040] [General]
I'm currently exploring the MaterialBuilder code.

As far as I understand if there're 2 objects, let's say cubes, one cube is affected by the directional light and the second one is not.

So, 
1st cube needs `DIR` variant
2nd cube does not `DIR` variant

Thus 4 separate shaders are needed. vertex + fragment shader with a directional light code and vertex + fragment shader without a directional light code.

As a point of reference, I use `gltf_viewer` example.

`JitShaderProvider::getMaterial()` function calls `JitShaderProvider::createMaterial()` which instantiates `MaterialBuilder builder` and generates shaders. `JitShaderProvider::createMaterial()` is called only once, so I thought it only generates vertex + fragment shaders.

But no! When I set the breakpoint inside `MaterialBuilder::generateShaders` function, count of `const std::vector<Variant>& variants` array is 42 which means there're 42 shaders generated only for one object. 

![image](https://github.com/google/filament/assets/43299401/ab26230c-e413-42b4-8b55-e8be32aa4d09)

Why there are so many shaders generated?

**Comment by romainguy** [comment]
The `MaterialProvider` (and `MaterialBuilder`) cannot know ahead of time what variants are needed (how would `gltfio` know whether you are going to use a directional light or shadows for instance?). This means all possible variants are *generated*, but *not compiled*. Filament will select the appropriate generated variant and compile only what's needed when it's needed.

**Comment by SC5Shout** [comment]
Alright, it makes sense. The `View` class controls light variants etc. + the material model filters these as needed. Thanks
